lib/Driver/ToolChains/Flang.cpp

```diff
@@ -600,27 +618,49 @@ void FlangFrontend::ConstructJob(Compilation &C, const JobAction &JA,
   // Add system include arguments.
   getToolChain().AddFlangSystemIncludeArgs(Args, UpperCmdArgs);
 
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("unix");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__unix");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__unix__");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("linux");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__linux");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__linux__");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__NO_MATH_INLINES");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__LP64__");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__LONG_MAX__=9223372036854775807L");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__SIZE_TYPE__=unsigned long int");
-  UpperCmdArgs.push_back("-def"); UpperCmdArgs.push_back("__PTRDIFF_TYPE__=long int");
+  // Use clang's predefined macros
+  DiagnosticsEngine DE(new DiagnosticIDs(), new DiagnosticOptions, new IgnoringDiagConsumer());
+  std::shared_ptr<clang::TargetOptions> TO = std::make_shared<clang::TargetOptions>();
+  TO->Triple = getToolChain().getEffectiveTriple().getTriple();
+  std::shared_ptr<TargetInfo> TI(clang::TargetInfo::CreateTargetInfo(DE, TO));
+  std::string PredefineBuffer;
+  llvm::raw_string_ostream Predefines(PredefineBuffer);
+  FlangMacroBuilder Builder(UpperCmdArgs, Args, Predefines);
+
+  LangOptions LO;
+  VersionTuple VT = getToolChain().computeMSVCVersion(&getToolChain().getDriver(), Args);
+  if (!VT.empty()) {
+    LO.MSCompatibilityVersion = VT.getMajor() * 10000000 +
```

Comment:
I see that this expression mirrors the expression in `lib/Frontend/CompilerInvocation.cpp`. Can you just add a comment about this expression though? Those seemingly random constants might be confusing.

Reply:
Done in 3f8fd4b